src/middleware/devtools.ts

```diff
@@ -44,7 +44,10 @@ type StoreDevtools<S> = S extends {
 }
   ? {
       setState(
-        ...a: [...a: TakeTwo<A>, actionType?: string | { type: unknown }]
+        ...a: [
+          ...a: TakeTwo<A>,
+          actionType?: string | { type: unknown; payload?: unknown }
+        ]
       ): Sr
```

Comment:
I think the problem is that the following does not compile...

```ts
import create from "zustand"
import { devtools } from "zustand/middleware"

interface BearState {
  bears: number
  increase: (by: number) => void
}

const useBearStore = create<BearState>()(devtools((set) => ({
  bears: 0,
  increase: by => set(
    (state) => ({ bears: state.bears + by }),
    false,
    { type: "increase", by }
//                      ~~[1]
  ),
})))

// [1]:
// Argument of type '{ type: string; by: number; }' is not assignable to parameter of type 'string | { type: unknown; } | undefined'.
//   Object literal may only specify known properties, and 'by' does not exist in type '{ type: unknown; }'.
```

This is because of TypeScript's "excess property checks", the correct way to fix is to tell typescript that the type of `action` is not `{ type: unknown }` but rather `extends { type: unknown }` like so...

```suggestion
      setState<A extends string | { type: unknown }>(
        ...a: [...a: TakeTwo<A>, actionType?: string | { type: unknown }]
        ...a: [
          ...a: TakeTwo<A>,
          action?: A
        ]
      ): Sr
```

Another way to fix like this...

```ts
setState(
  ...a: [...a: TakeTwo<A>, actionType?: string | { type: unknown }]
  ...a: [
    ...a: TakeTwo<A>,
    action?: string | { type: unknown } & { [_ in string]: unknown }
  ]
): Sr
```

But this seems a little less correct and in general it's good to stay away from index signatures (ie { [_ in string]: unknown }).

And the current fix `{ type: unknown, payload?: unknown }` doesn't make the cut because it doesn't really solve the actual problem showed above, `{ type: "increase", by }` should also be assignable to `action` parameter.

Reply:
Try `{ type: "increase", payload: by }`

Reply:
I don't follow, `{ type: "increase", by }` should compile, ofc users can write `{ type: "increase", payload: by }` if they wish.

Reply:
I think this conversation is connected with the other one from this PR. Let's continue here: https://github.com/pmndrs/zustand/pull/1183#discussion_r943434166.